ksm.go

```diff
@@ -0,0 +1,214 @@
+//
+// Copyright (c) 2017 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+)
+
+type ksmSetting struct {
+	pagesPerScanFactor uint32
+	scanIntervalMs     uint32
+}
+
+const (
+	ksmSlug       = "slug"
+	ksmSlow       = "slow"
+	ksmStandard   = "standard"
+	ksmAggressive = "aggressive"
+)
+
+var ksmSettings = map[string]ksmSetting{
+	ksmSlug:       {1000, 500}, // Every 500ms, we scan 1 page for every 1000 pages available in the system
+	ksmSlow:       {500, 100},  // Every 100ms, we scan 1 page for every 500 pages available in the system
+	ksmStandard:   {100, 10},   // Every 10ms, we scan 1 page for every 100 pages available in the system
+	ksmAggressive: {10, 1},     // Every ms, we scan 1 page for every 10 pages available in the system
+}
+
+const (
+	ksmRoot           = "/sys/kernel/mm/ksm/"
+	ksmRunFile        = "run"
+	ksmPagesToScan    = "pages_to_scan"
+	ksmSleepMillisec  = "sleep_millisecs"
+	ksmRun            = "1"
+	defaultKSMSetting = ksmAggressive
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+
```

Comment:
nitpick: this func has a blank line before the return, but the near mirror func above it does not - just caught my eye.